### PR TITLE
RAF: extract more details when printing structure

### PR DESCRIPTION
**Pull request by kmilos**
Going by https://libopenraw.freedesktop.org/formats/raf/ some bytes after the first meta and CFA offset and length were unknown. @hfiguiere

However, I think these are to describe the file layout when the RAF has two sub-images (like the ones from EXR sensors), so I posit it is a specification of some interleaving scheme w/ skip? (haven't seen anything other than 0), chunk size, and stride....

For the FinePix F700 sample from RPU, this now outputs:
```
STRUCTURE OF RAF FILE: Fujifilm - FinePix F700 - double width 16-bit unpacked little-endian (4_3).RAF
   Address |    Length | Payload
         0 |        16 |        magic : FUJIFILMCCD-RAW
        16 |         4 |        data1 : 0201
        20 |         8 |        data2 : FF390101
        28 |        32 |       camera : FinePix F700
        60 |         4 |      version : 0264
        64 |        20 |      unknown : ...................
        84 |         4 |  JPEG offset : 148
        88 |         4 |  JPEG length : 482850
        92 |         4 | meta offset1 : 483000
        96 |         4 | meta length1 : 2148
       100 |         4 |  CFA offset1 : 487296
       104 |         4 |  CFA length1 : 6382592
       108 |         4 |    CFA skip1 : 0
       112 |         4 |   CFA chunk1 : 2944
       116 |         4 |  CFA stride1 : 5888
       120 |         4 | meta offset2 : 485148
       124 |         4 | meta length2 : 2148
       128 |         4 |  CFA offset2 : 490240
       132 |         4 |  CFA length2 : 6382592
       136 |         4 |    CFA skip2 : 0
       140 |         4 |   CFA chunk2 : 2944
       144 |         4 |  CFA stride2 : 5888
       148 |    482850 |    JPEG data : ....%"Exif..II*
    483000 |      2148 |   meta data1 : ...).....x......
    485148 |      2148 |   meta data2 : ...).....x......
    487296 |   6382592 |    CFA data1 : ...............
    490240 |   6382592 |    CFA data2 : ...............
```
Note how the offset of the 2nd CFA data payload is one CFA chunk size of 2944 after the 1st one, giving the hint about interleaved chunks.

NB: "CFA length" seems to be the effective one, after deinterleaving.

### CODE REVIEW

File: src/rafimage.cpp [hunk 1]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
 
     long jpg_img_off = Exiv2::getULong(jpg_img_offset, bigEndian);
     long jpg_img_len = Exiv2::getULong(jpg_img_length, bigEndian);
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:141]:
```suggestion
    uint32_t jpg_img_off = Exiv2::getULong(jpg_img_offset, bigEndian);
    uint32_t jpg_img_len = Exiv2::getULong(jpg_img_length, bigEndian);
```

File: src/rafimage.cpp [hunk 2]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
-    c_off << cfa_off;
-    c_len << cfa_len;
-    {
-      out << Internal::indent(depth) << Internal::stringFormat(format, address, 4L) << "TIFF Offset : " << c_off.str()
-          << std::endl;
-      out << Internal::indent(depth) << Internal::stringFormat(format, address2, 4L) << "TIFF Length : " << c_len.str()
-          << std::endl;
+    // RAFs can carry the payload in one or two parts
+    long meta_off[2], meta_len[2];
+    long cfa_off[2], cfa_len[2], cfa_skip[2], cfa_size[2], cfa_stride[2];
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:155]:
```suggestion
    uint32_t meta_off[2], meta_len[2];
    uint32_t cfa_off[2], cfa_len[2], cfa_skip[2], cfa_size[2], cfa_stride[2];
```

File: src/rafimage.cpp [hunk 3]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+    for (size_t i = 0; i < 2; i++) {
+      address = io_->tell();
+      byte data[4];
+      io_->read(data, 4);
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:159]:
```suggestion
      io_->readOrThrow(data, 4);
```

File: src/rafimage.cpp [hunk 4]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+      meta_off[i] = Exiv2::getULong(data, bigEndian);
+      address2 = io_->tell();
+      io_->read(data, 4);
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:162]:
```suggestion
      io_->readOrThrow(data, 4);
```

File: src/rafimage.cpp [hunk 5]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+            << c_off.str() << std::endl;
+        out << Internal::indent(depth) << Internal::stringFormat(format, address2, 4L) << "meta length" << i + 1
+            << " : " << c_len.str() << std::endl;
+      }
+
+      address = io_->tell();
+      io_->read(data, 4);
+      cfa_off[i] = Exiv2::getULong(data, bigEndian);
+      address2 = io_->tell();
+      io_->read(data, 4);
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:179]:
```suggestion
      io_->readOrThrow(data, 4);
```

File: src/rafimage.cpp [hunk 6]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+      cfa_len[i] = Exiv2::getULong(data, bigEndian);
+      size_t address3 = io_->tell();
+      io_->read(data, 4);
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:182]:
```suggestion
      io_->readOrThrow(data, 4);
```

File: src/rafimage.cpp [hunk 7]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+      cfa_skip[i] = Exiv2::getULong(data, bigEndian);
+      size_t address4 = io_->tell();
+      io_->read(data, 4);
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:185]:
```suggestion
      io_->readOrThrow(data, 4);
```

File: src/rafimage.cpp [hunk 8]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+      cfa_size[i] = Exiv2::getULong(data, bigEndian);
+      size_t address5 = io_->tell();
+      io_->read(data, 4);
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:188]:
```suggestion
      io_->readOrThrow(data, 4);
```

File: src/rafimage.cpp [hunk 9]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+            << " : " << c_size.str() << std::endl;
+        out << Internal::indent(depth) << Internal::stringFormat(format, address5, 4L) << " CFA stride" << i + 1
+            << " : " << c_stride.str() << std::endl;
+      }
     }
 
     io_->seek(jpg_img_off, BasicIo::beg);  // rewind
     address = io_->tell();
     DataBuf payload(16);  // header is different from chunks
     io_->read(payload.data(), payload.size());
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:217]:
```suggestion
    io_->readOrThrow(payload.data(), payload.size());
```

File: src/rafimage.cpp [hunk 10]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
-      out << Internal::indent(depth) << Internal::stringFormat(format, address, jpg_img_len)  // , jpg_img_off)
-          << "       JPEG : " << Internal::binaryToString(makeSlice(payload, 0, payload.size())) << std::endl;
+      out << Internal::indent(depth) << Internal::stringFormat(format, address, jpg_img_len)
+          << "   JPEG data : " << Internal::binaryToString(makeSlice(payload, 0, payload.size())) << std::endl;
     }
 
-    io_->seek(cfa_hdr_off, BasicIo::beg);  // rewind
+    io_->seek(meta_off[0], BasicIo::beg);  // rewind
     address = io_->tell();
     io_->read(payload.data(), payload.size());
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:225]:
```suggestion
    io_->readOrThrow(payload.data(), payload.size());
```

File: src/rafimage.cpp [hunk 11]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
-          << "        CFA : " << Internal::binaryToString(makeSlice(payload, 0, payload.size())) << std::endl;
+      out << Internal::indent(depth) << Internal::stringFormat(format, address, meta_len[0])
+          << "  meta data1 : " << Internal::binaryToString(makeSlice(payload, 0, payload.size())) << std::endl;
     }
 
-    io_->seek(cfa_off, BasicIo::beg);  // rewind
+    if (meta_off[1] && meta_len[1]) {
+      io_->seek(meta_off[1], BasicIo::beg);  // rewind
+      address = io_->tell();
+      io_->read(payload.data(), payload.size());
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:234]:
```suggestion
      io_->readOrThrow(payload.data(), payload.size());
```

File: src/rafimage.cpp [hunk 12]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+      {
+        out << Internal::indent(depth) << Internal::stringFormat(format, address, meta_len[1])
+            << "  meta data2 : " << Internal::binaryToString(makeSlice(payload, 0, payload.size())) << std::endl;
+      }
+    }
+
+    io_->seek(cfa_off[0], BasicIo::beg);  // rewind
     address = io_->tell();
     io_->read(payload.data(), payload.size());
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:243]:
```suggestion
    io_->readOrThrow(payload.data(), payload.size());
```

File: src/rafimage.cpp [hunk 13]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
-      out << Internal::indent(depth) << Internal::stringFormat(format, address, cfa_len)  // cfa_off
-          << "       TIFF : " << Internal::binaryToString(makeSlice(payload, 0, payload.size())) << std::endl;
+      out << Internal::indent(depth) << Internal::stringFormat(format, address, cfa_len[0])
+          << "   CFA data1 : " << Internal::binaryToString(makeSlice(payload, 0, payload.size())) << std::endl;
+    }
+
+    if (cfa_off[1] && cfa_len[1]) {
+      io_->seek(cfa_off[1], BasicIo::beg);  // rewind
+      address = io_->tell();
+      io_->read(payload.data(), payload.size());
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:252]:
```suggestion
      io_->readOrThrow(payload.data(), payload.size());
```

File: src/rafimage.cpp [hunk 14]
```diff
@@ -71,19 +71,19 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+    constexpr auto format = " %9zu | %9ld | ";
 
     {
       out << Internal::indent(depth) << "STRUCTURE OF RAF FILE: " << io().path() << std::endl;
-      out << Internal::indent(depth) << "  Address |   Length | Payload" << std::endl;
+      out << Internal::indent(depth) << "   Address |    Length | Payload" << std::endl;
     }
 
     byte magicdata[17];
     io_->read(magicdata, 16);
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:82]:
```suggestion
    io_->readOrThrow(magicdata, 16);
```

File: src/rafimage.cpp [hunk 15]
```diff
@@ -119,15 +119,15 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
     dir_version[4] = 0;
     {
       out << Internal::indent(depth) << Internal::stringFormat(format, address, 4L)  // 60
-          << "    version : " << std::string(reinterpret_cast<char*>(&dir_version)) << std::endl;
+          << "     version : " << std::string(reinterpret_cast<char*>(&dir_version)) << std::endl;
     }
 
     address = io_->tell();
     DataBuf unknown(20);
     io_->read(unknown.data(), unknown.size());
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:127]:
```suggestion
    io_->readOrThrow(unknown.data(), unknown.size());
```

File: src/rafimage.cpp [hunk 16]
```diff
@@ -139,78 +139,121 @@ void RafImage::printStructure(std::ostream& out, PrintStructureOption option, si
...
+        c_off << meta_off[i];
+        c_len << meta_len[i];
+        out << Internal::indent(depth) << Internal::stringFormat(format, address, 4L) << "meta offset" << i + 1 << " : "
+            << c_off.str() << std::endl;
+        out << Internal::indent(depth) << Internal::stringFormat(format, address2, 4L) << "meta length" << i + 1
+            << " : " << c_len.str() << std::endl;
+      }
+
+      address = io_->tell();
+      io_->read(data, 4);
```
Comment (kevinbackhouse) [inline src/rafimage.cpp:176]:
```suggestion
      io_->readOrThrow(data, 4);
```

**Comment by kmilos** [comment]
Thanks for all the changes @kevinbackhouse

**Review by kevinbackhouse — APPROVED** [review]

**Pull request merged**